### ISSUE TITLE
Method to limit depth of layer blending

### ISSUE THREAD

**Issue by schreyack** [xLightsSequencer/xLights#4552]
I'm always frustrated when I want to apply a layer blending mask in layer 1 which impacts layer 2, but not layer3.

I would like to be able to apply layer blending of "Blend End" to layer 3, which would then allow all layers below (layers 4+) to not have any impact from the blending in layer 1.

I have tried every method I can think of to get around the mask, such as applying the mask to a different group, but there doesn't seem to be any way to avoid the mask.  If there is already a way to do this, I apologize.

**Comment by computergeek1507** [comment]
I'm pretty sure you can do this with Canvas mode and off effects

**Comment by schreyack** [comment]
I did try that... layer 3 is off effect (i tried both with transparent and not) set to normal and canvas mode.  layer 4 is on effect  and selected as the layer to blend into layer 3.  I also tried layer 3 as on effect.  In all cases, nothing overrides the mask from layer 1.

**Comment by cybercop23** [comment]
Can you create a sample layout with one/two models and a sample sequence with the setting(s) you have tried?

**Comment by schreyack** [comment]
> Is layer 1 the very top layer or the bottom layer? Cause the bottom layer renders first and the top layer renders last so you won't override the top layer using layers below it right?

No, layers render top to bottom, unlike elements which render bottom to top.

**Comment by schreyack** [comment]
> Can you create a sample layout with one/two models and a sample sequence with the setting(s) you have tried?

Here you go!
[Test.zip](https://github.com/xLightsSequencer/xLights/files/15241553/Test.zip)

**Comment by dkulp** [comment]
What are you expecting it to do?  It looks correct to me.   Layer 2 is a white on effect with "normal" blending.   Thus, everything below that is pretty much irrelevant.    Layer 1 is Circles as unmask which would be unmasking the white from the layer below it.

**Comment by schreyack** [comment]
> What are you expecting it to do? It looks correct to me. Layer 2 is a white on effect with "normal" blending. Thus, everything below that is pretty much irrelevant. Layer 1 is Circles as unmask which would be unmasking the white from the layer below it.

this is a somewhat silly example - but i want the layer 1 unmask to apply to layer 2 only, so that i can then apply an effect in layer 3+ that would show behind the circles being unmasked by layer 1.

**Comment by schreyack** [comment]
here's an example that hopefully explains better what i'm trying to accomplish - i'm using the layer 1 unmask to create an animation of a rocket trail (think 4th of july show) moving across the scene by only showing a bit of a pinwheel in layer 2.  which works great.  but then i want to have an effect like twinkle in layer 3 showing on the entire background, not just the area being exposed by the unmask.  does that make sense?
[Test 2.zip](https://github.com/xLightsSequencer/xLights/files/15242098/Test.2.zip)

**Comment by dkulp** [comment]
The question to me is whether the bg.Set(0, 0, 0); in the unmasks should be bg.Set(0, 0, 0, 0); instead.   I need to experiment a bit more with that to see what it may break.   That would allow a canvas to be used for the pinwheel+shape that would then have the "black" areas be transparent and not black.

**Comment by MrPierreB** [comment]
You can do what you want already by either re-ordering your layers to achieve the effect or you can use canvas mode and using off effects to apply the layers below above the masking your are doing.

For example in your example.  you can place the twinkle effect as the top layer and have the layer blending set to layered.  I know that is beside the point here as you want to know how to do it if it was at the bottom.
The way you could do it when the layer is below the masking effect as you suggest.. is to have the layers in this case.
The very bottom layer is your twinkle effect. Then you have an standard off effect on the layer above it to turn it off so it is not shown on the pinwheel layer which is the layer above it.  then above the pinwheel you have your shape effect with your 1 is true unmask layer blending style.  then the top layer is another off effect, but this one you set to Canvas mode and you click the layers button and select Layer 5. Set it to be transparent and then set your layer blending to Layered.
That will accomplish the same thing.

If you have multiple layers and not only a twinkle you want to show above, you will use another off effect to combine them using transparent and persistent.

**Comment by schreyack** [comment]
@MrPierreB thank you!  that works as you described.

@dkulp @AzGilrock given there is a solution to this problem (although, i would describe it as a work around) i don't know if it's worth the development effort to have a less convoluted way to achieve this.  but it would be nice. :)

**Comment by schreyack** [comment]
because you set the layer blending to "layered".

**Comment by dkulp** [comment]
I just pushed a bunch of changes that will make this a bit easier.    The main problem was due to "Canvas" mode effects always ending up being 100% solid (alpha always at 255) which then didn't allow any decent way of blending with layers below it that were not part of the canvas.   I fixed that, but that doesn't really address this as the "True Unmask" blending also sets everything to solid black.    To deal with that, I've added new "Styles" to the transparent off effect used for canvas blending to provide options to convert Black -> Transparent.     This allows a Canvas off to blend the two layers to create the meteor, but making all the "Black" outside the meteor to be transparent which would then blend onto the twinkles below it like a normal effect.

**Comment by schreyack** [comment]
wow, that was quick!  thank you!!

**Comment by MrPierreB** [comment]
I always understood it as on mulitple layers it renders from the bottom up.. it would apply the layer above to the layer below, and after it did its processing , that now becomes 'Effect 2' and will render the layer above onto that , then that becomes 'Effect 2' etc...
And canvas mode is taking the layers below and drawing it physically onto that layer as if the pixels were on that layer (and you can select which layers it uses).  So in this sense it makes sense and working from the bottom up works.   Don't think going other way and now limitting how many layers down it affects would be wise. 
Would definetly break those who use many layers (I'm releasing a sequence soon and hoping this change wont affect it) and I know Doug Reel uses alot of canvas mode.
